Use docker to run monerod and wallet

tcharding opened the pull request:
Use docker along with testcontainers to run a xmr.to docker image. Start
an instance of `monerod` in regtest mode and also start an instance of
`monero-wallet-rpc`.

Add functions to initialise the container how we would like it, this means

- Create a wallet
- Initialise accounts for Alice and Bob incl. fund Alice 1000 moneoj.
- Kick of mining blocks every second.

Commit:
Use docker to run monerod and wallet

Use docker along with testcontainers to run a xmr.to docker image. Start
an instance of `monerod` in regtest mode and also start an instance of
`monero-wallet-rpc`.

Add functions to initialise the container how we would like it, this means

- Create a wallet
- Initialise accounts for Alice and Bob incl. fund Alice 1000 moneoj.
- Kick of mining blocks every second.

bonomat committed:
Tmp commit

Commit:
current

bonomat committed:
Start monerod and assert blockheight (this test should fail if no additional block was generated)

Commit:
Remove commented out code

bonomat committed:
Wait for monerod to start

bonomat committed:
Remove dead test

Commit:
Merge branch 'genericdocker' of github.com:coblox/monero-harness-rs into genericdocker

bonomat committed:
Provide new function for starting monerod and monerowalletrpc

Commit:
Merge branch 'genericdocker' of github.com:coblox/monero-harness-rs into genericdocker

Commit:
Add integration tests

Commit:
Remove password from wallet instance

Commit:
Fix accounts integration test

bonomat committed:
Add mining and init function

Commit:
Add debugging output

Commit:
wip: mining and init are buggy

Commit:
Some fixes but still buggy

bonomat commented on src/lib.rs:
```diff
@@ -17,46 +17,151 @@
+    /// Initialise by creating a wallet, generating some `blocks`, and starting
+    /// a miner thread that mines to the primary account. Also create two
+    /// sub-accounts, one for Alice and one for Bob. Fund Alice's account.
+    pub async fn init_with_accounts(&self) -> Result<()> {
```
I think we should pass-in the funding amounts as parameters instead of constants.

bonomat commented on src/lib.rs:
```diff
@@ -17,46 +17,151 @@
+
+        let miner = self.wallet.get_address_primary().await?.address;
+
+        let res = self.monerod.generate_blocks(70, miner.clone()).await?;
```
I'd add a note why 70 blocks. According to the specs it should be 60 but it seems only to work with 70 blocks.
```suggestion
		// 70 blocks seems to be the sweet number to have enough spendable funding.
        let res = self.monerod.generate_blocks(70, miner.clone()).await?;
```

bonomat committed:
A bit of cleanup

bonomat commented on src/lib.rs:
```diff
@@ -17,46 +17,151 @@
+    // It takes a little while for the wallet to sync with monerod.
+    async fn wait_for_wallet_block_height(&self, height: u32) -> Result<()> {
+        while self.wallet.block_height().await?.height < height {
+            tokio::time::delay_for(Duration::from_millis(WAIT_WALLET_SYNC_MILLIS)).await;
```
🚧 
`WAIT_WALLET_SYNC_MILLIS` is not defined!

bonomat commented on src/lib.rs:
```diff
@@ -17,46 +17,151 @@
+
+/// When creating an account for Alice fund it with this amount of moneroj.
+const INITIAL_FUNDS_ALICE: u64 = 1000;
+/// How often we mine a block.
+const BLOCK_TIME_SECS: u32 = 1;
+/// Poll interval when checking if the wallet has synced with monerod.
+const BLOCK_TIME_SECS: u32 = 200;
```
🚧
Duplicated!

bonomat replied:
🚧
type should be `u64` as `Duration::from_secs(xx)` takes  `u64`

bonomat reviewed (changes requested):
Did you run this locally? 
We should add a CI soon to catch errors early.

bonomat commented:
I'll resolve the mentioned issues and push to this PR.

bonomat committed:
Make init accounts more configurable for Alice and Bob

bonomat commented:
@tcharding : I approved (my own changes). Please merge at your convenience :)

// edit // I merged master into this PR to enable CI

bonomat committed:
Merge remote-tracking branch 'origin/master' into merge

D4nte commented on Cargo.toml:
```diff
@@ -7,9 +7,14 @@ edition = "2018"
-url = "2"
\ No newline at end of file
+url = "2"
+
```
double line at the end

tcharding replied:
Cheers :)

D4nte commented on src/monerod.rs:
```diff
@@ -4,8 +4,14 @@ use anyhow::Result;
 use reqwest::Url;
 use serde::{Deserialize, Serialize};
 
+// #[cfg(not(test))]
```
Comments left

Commit:
Use &str for parameter arguments

Use `String` for arguments forces the caller to clone, this is unnecessary,
instead put the burned on the function and accept a reference then call
`to_owned` as necessary.

Commit:
Use zero to signal no funding instead of None

We cannot do a transfer of 0, we can however conditionally transfer based on
zero value instead of using Some/None. This makes the call site more intuitive.

Commit:
Remove end of file whitespace

thomaseizinger commented on src/lib.rs:
```diff
@@ -17,46 +17,171 @@
 #[derive(Debug)]
-pub struct Client {
+pub struct Client<'c> {
+    container: Container<'c, clients::Cli, GenericImage>,
```
It is a bit odd that this struct is called _Client_ if it hosts the container that runs the actual node. Seems like it might be worth separating out an actual node and let this struct host the two clients.

tcharding replied:
How about just renaming it? If we pull the container out its just another thing that users have to instantiate and keep around, they already have to create the docker client.

thomaseizinger commented on src/lib.rs:
```diff
@@ -17,46 +17,171 @@
+
+    /// Initialise by creating a wallet, generating some `blocks`, and starting
+    /// a miner thread that mines to the primary account.
+    pub async fn init(&self, blocks: u32) -> Result<()> {
```
Having multiple `init` methods that all just take `&self` doesn't make use of the type system really well.

I'd assume you would only want one of those to actually be called. Could make multiple constructors instead so one has to choose which one to run.

tcharding replied:
Good idea, thanks.

tcharding replied:
Ok, I messed around with this for ages - way too long. In the end I removed the second constructor, removed one of the inits and commented the other saying its probably not the one you want (kept it because it is useful for testing but `#[cfg(test)] doesn't help since we test it in `/tests/`. @bonomat I removed the default init that I argue for having in there :)

thomaseizinger commented on src/wallet.rs:
```diff
@@ -4,6 +4,15 @@ use anyhow::Result;
+// TODO: Either use println! directly or import tracing also?
+use std::println as debug;
+
+// TODO: Consider using bignum for moneroj instead of u64?
```
FYI: Bignum is not copy which can be annoying.

tcharding replied:
Noted, cheers.

thomaseizinger commented on tests/rpc.rs:
```diff
@@ -0,0 +1,121 @@
+#[tokio::test]
+async fn connect_to_monerod() {
+    let docker = clients::Cli::default();
+    let cli = Client::new_with_random_local_ports(&docker);
```
Stale variable name :)

tcharding replied:
I don't understand sorry. Is this referencing your comment above about the name of this type?

thomaseizinger replied:
Yes

thomaseizinger commented on tests/rpc.rs:
```diff
@@ -0,0 +1,121 @@
+
+#[tokio::test]
+async fn miner_is_running_and_producing_blocks() {
+    let docker_client = clients::Cli::default();
```
Just a heads-up: `testcontainers` is not built for async contexts as it blocks the thread for every operation. Just be mindful of that because it means you will block the executor thread and not just the single future you are using it in.

tcharding replied:
Oh man that's a big fail, since the whole point of this crate is to enable tests that each spin up a container to run in parallel. I guess we still get the advantage of clean/separate state in each container even if the tests end up being run in series.

bonomat replied:
Imho `to run in parallel` is not that important. The library is more about having a clean test environment for our protocols.

Commit:
Remove bignum consideration comment

tcharding commented:
Thanks for the review @thomaseizinger

Commit:
Re-order dependency list

Commit:
Refactor out function spin_up_container

Commit:
Move random port constructor up a bit

Commit:
Remove default fund initialiser

Commit:
Remove new()

Commit:
Simplify create/init

Commit:
Remove --features test-docker from CI
